### PR TITLE
Add Arrow and ArrowObject

### PR DESCRIPTION
**Pull request by Licini**
Adding the `Arrow` under `shapes` folder (which can be migrated to compas later) and corresponding  `ArrowObject`.
testing script: `scripts/v120_arrows.py`

image

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by Licini** [comment]
Performance-wise, It runs without significant delay on my pc and mac with around 1000 arrows, it we are to deal with more demanding situation later, we can either make an arrow group object or go for shader-based arrows

File: src/compas_view2/shapes/arrow.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    # ==========================================================================
+
+    def to_vertices_and_faces(self, u=3):
+        """Returns a list of vertices and faces.
+
+        Parameters
+        ----------
+        u : int, optional
+            Number of faces in the "u" direction.
+            Default is ``10``.
```
Comment (brgcode) [inline src/compas_view2/shapes/arrow.py:119]: default is currently `3`.
i would set the default to `16`...

File: src/compas_view2/shapes/arrow.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    ----------
+    position : list of float or `compas.geometry.Vector`
+        The start position of arrow
+    direction : list of float or `compas.geometry.Vector`
+        The height of the cone.
+
+    Attributes
+    ----------
+    head_portion : float
+        The portion of head of the arrow
```
Comment (brgcode) [inline src/compas_view2/shapes/arrow.py:29]: weird name :)
is this a percentage of the total length?

File: src/compas_view2/shapes/arrow.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+from compas.geometry import Circle
+from compas.geometry import Cylinder
+from compas.geometry import Plane
+from compas.geometry import Shape
+from compas.datastructures import Mesh
+
+__all__ = ['Arrow']
+
+
+class Arrow(Shape):
```
Comment (brgcode) [inline src/compas_view2/shapes/arrow.py:16]: should this not be a PR for `compas.geometry`?

File: src/compas_view2/shapes/arrow.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        cone = Cone(circle, self.direction.length*self.head_portion)
+        v, f = cone.to_vertices_and_faces(u=u)
+        head = Mesh.from_vertices_and_faces(v, f)
+
+        # COnstruct the body of arrow
+        body_center = self.position + self.direction*(1-self.head_portion)/2
+        plane = Plane(body_center, self.direction)
+        circle = Circle(plane, self.body_width*self.direction.length)
+        cylinder = Cylinder(circle, self.direction.length *
+                            (1 - self.head_portion))
```
Comment (brgcode) [inline src/compas_view2/shapes/arrow.py:143]: if you modify the line length setting for flake8, this is not necessary...

File: src/compas_view2/objects/__init__.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 from compas.geometry import Box
 from compas.geometry import Sphere
 from compas.geometry import Torus
+from ..shapes import Arrow
```
Comment (brgcode) [inline src/compas_view2/objects/__init__.py:8]: move to `compas.geometry`?

File: src/compas_view2/shapes/arrow.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        The portion of head of the arrow
+    head_width : float
+        The head width relative to the length of arrow
+    body_width : float
+        The body width relative to the length of arrow
+
+
+    Examples
+    --------
+    >>> from compas.geometry import Arrow
```
Comment (GeneKao) [inline src/compas_view2/shapes/arrow.py:38]: is Arrow in `compas.geometry` or should it be part of view2 package?

**Comment by GeneKao** [comment]
just curious, is opengl have its native arrow object? or it has to be a Mesh like you currently implemented?

**Comment by brgcode** [comment]
there is no arrow object. it has to be a mesh if the arrow is "3d". alternatively, we draw lines with a "2d" arrowhead directly on the screen...

**Pull request merged**